mono/utils/mono-threads.c

```diff
@@ -993,37 +1004,104 @@ mono_thread_info_resume (MonoNativeThreadId tid)
 	return result;
 }
 
-gboolean
-mono_thread_info_begin_suspend (MonoThreadInfo *info)
+static MonoThreadBeginSuspendResult
+begin_suspend_phase_coop (MonoThreadInfo *info);
+
+static MonoThreadBeginSuspendResult
+begin_suspend_phase_all (MonoThreadInfo *info);
+
+
+MonoThreadBeginSuspendResult
+mono_thread_info_begin_suspend (MonoThreadInfo *info, MonoThreadSuspendPhase phase)
+{
+	/* FIXME: rationalize this spaghetti */
+	if (phase == MONO_THREAD_SUSPEND_PHASE_ALL && mono_threads_is_hybrid_suspension_enabled ())
```

Comment:
Won't the `mono_threads_is_hybrid_suspension_enabled` check preclude preempt suspend to be used?

Reply:
The phase names and the helper function names are bad.

`begin_suspend_phase_coop`:
* in full preemptive it will do preemptive suspend of all threads
* in full coop it will do cooperative suspend of all threads (treating blocking threads as logically suspended).
* in hybrid suspend it will do cooperative suspend of running threads and it will change the thread state to request a suspend for blocking threads (BLOCKING -> BLOCKING_SUSPEND_REQUESTED), but it won't initiate any suspend mechanism yet and will return MONO_THRREAD_BEGIN_SUSPEND_NEXT_PHASE.

Reply:
Now it makes sense. Yes, naming is bad.
I guess once we find good names for the phases, this will be easily fixable to be `begin_suspend_XXX_phase`